src/py/flwr/server/superlink/state/in_memory_state.py

```diff
@@ -134,6 +134,28 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
                 )
                 return None
 
+            # Fail if the TaskRes TTL exceeds the
+            # expiration time of the TaskIns it replies to.
+            # Condition: TaskIns.created_at + TaskIns.ttl ≥
+            #            TaskRes.created_at + TaskRes.ttl
+            # A small tolerance is introduced to account
+            # for floating-point precision issues.
+            max_allowed_ttl = (
+                task_ins.task.created_at + task_ins.task.ttl - task_res.task.created_at
+            )
+            if (
+                task_res.task.ttl
+                and (task_res.task.ttl - max_allowed_ttl) > TTL_TOLERANCE
+            ):
+                log(
+                    ERROR,
+                    "Received TaskRes with TTL %.2f "
+                    "exceeding the allowed maximum TTL %.2f.",
+                    task_res.task.ttl,
+                    max_allowed_ttl,
+                )
+                return None
```

Comment:
This block of code and the one in `sqlite_state` should live in a shared function to reduce code duplication.